packages/blockchain-utils-linking/src/ethereum.ts

```diff
@@ -0,0 +1,189 @@
+  );
+}
+
+export function normalizeAccountId(account: AccountID): AccountID {
+  account.address = account.address.toLowerCase();
+  return account;
+}
+
+function utf8toHex(message: string): string {
+  const bytes = uint8arrays.fromString(message);
+  const hex = uint8arrays.toString(bytes, "base16");
+  return "0x" + hex;
+}
+
+async function createEthLink(
+    did: string,
+    account: AccountID,
+    provider: any,
+    opts: any = {}
+): Promise<LinkProof> {
+  const { message, timestamp } = getConsentMessage(did, !opts.skipTimestamp);
+  const hexMessage = utf8toHex(message);
+  const payload = encodeRpcMessage("personal_sign", [
+    hexMessage,
+    account.address,
+  ]);
+  const signature = await safeSend(payload, provider);
+  const proof: LinkProof = {
+    version: 2,
+    type: ADDRESS_TYPES.ethereumEOA,
+    message,
+    signature,
+    account: account.toString(),
+  };
+  if (!opts.skipTimestamp) proof.timestamp = timestamp;
+  return proof;
+}
+
+async function validateChainId(
+    account: AccountID,
+    provider: any
+): Promise<void> {
+  const payload = encodeRpcMessage("eth_chainId", []);
+  const chainIdHex = await safeSend(payload, provider);
+  const chainId = parseInt(chainIdHex, 16);
+  if (chainId !== parseInt(account.chainId.reference)) {
+    throw new Error(
+        `ChainId in provider (${chainId}) is different from AccountID (${account.chainId.reference})`
+    );
+  }
+}
+
+async function createErc1271Link(
+    did: string,
+    account: AccountID,
+    provider: any,
+    opts: any
+): Promise<LinkProof> {
+  const ethLinkAccount = opts?.eoaSignAccount || account;
+  const res = await createEthLink(did, ethLinkAccount, provider, opts);
+  await validateChainId(account, provider);
+  return Object.assign(res, {
+    type: ADDRESS_TYPES.erc1271,
+    account: account.toString(),
+  });
+}
+
+export async function createLink(
+    did: string,
+    account: AccountID,
+    provider: any,
+    opts: any
+): Promise<LinkProof> {
+  account = normalizeAccountId(account);
+  if (await isERC1271(account, provider)) {
+    return createErc1271Link(did, account, provider, opts);
+  } else {
+    return createEthLink(did, account, provider, opts);
+  }
+}
+
+export async function authenticate(
+    message: string,
+    account: AccountID,
+    provider: any
+): Promise<string> {
+  if (account) account = normalizeAccountId(account);
+  if (provider.isAuthereum) return provider.signMessageWithSigningKey(message);
+  const hexMessage = utf8toHex(message);
+  const payload = encodeRpcMessage("personal_sign", [
+    hexMessage,
+    account.address,
+  ]);
+  const signature = await safeSend(payload, provider);
+  if (account) {
+    const recoveredAddr = verifyMessage(message, signature).toLowerCase();
+    if (account.address !== recoveredAddr)
+      throw new Error(
+          "Provider returned signature from different account than requested"
+      );
```

Comment:
Do we need this check really? If not we could remove the dependency on `@ethersproject/wallet` in this package.

Reply:
Really not. I consider it an attempt to be smart and preemptively fix a (probably non-existent) problem of a provider misbehaving.